### PR TITLE
LinearAlgebra.leastSquares()

### PR DESCRIPTION
**Pull request by ben-albrecht**
Add `LinearAlgebra.leastSquares` based on [`numpy.linalg.lstsq`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html#numpy.linalg.lstsq).

Also in this PR:

- Added tests for `leastSquares()`
- Fixed a bug in `LAPACK.gels{s,d,y}` interface where the `rank` output variable was not available as a ref arg.
- Added private machine epsilon helper functions to `LinearAlgebra` (work-around for https://github.com/chapel-lang/chapel/issues/14016)
- Refactored some of `LinearAlgebra/correctness/TestUtils.chpl`
- Added a `vander` function for generating Vandermonde matrices (useful for least squares user code)
  - Leaving undocumented for now, and will properly add it in a separate PR

Future work for this function:

- Support `overwrite=false` argument
- Support 2d `b` arrays
- Make `vander` function public

### CODE REVIEW

File: modules/packages/LinearAlgebra.chpl [hunk 1]
```diff
@@ -1270,57 +1269,58 @@ proc det (A: [?Adom] ?eltType) {
...
         y(j) -= U(j,i) * sol;
       }
     }
   }
-  
+
   return y;
 }
 
+//m x n * n => m
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1323]: Forgotten comment?

If not, it is not clear to me what you are referring to here.

File: modules/packages/LinearAlgebra.chpl [hunk 2]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+   ``cond`` is the cut-off threshold such that singular values will be
+   considered 0.0. If ``cond < 0.0`` (defaults to ``-1.0``), the treshold will
+   be set to ``max((...A.shape)) * epsilon``, where ``epsilon`` is the machine
+   precision for ``A.eltType``.
+
+   Returns a tuple of ``(x, residues, rank, s)``, where:
+
+  - ``x`` is the the least-squares solution with shape of ``b``
+  - ``residues`` is the square of the 2-norm for each column in ``b - a x`` if ``M > N`` and ``A.rank == n``
+    - ``residues`` is a scalar if ``b`` is 1-D, otherwise it is a 1-element array
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1346]: I am not sure how this bulleting would render, but something like this would be nice
```
- ``residues`` is;
  - square of ... if ....
  - a scalar if...
```

File: modules/packages/LinearAlgebra.chpl [hunk 3]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+  - ``rank`` is the effective rank of ``A``
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1347]: Ugh.. I wish `rank` wasn't overloaded :(

File: modules/packages/LinearAlgebra.chpl [hunk 4]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+  - ``s`` is the singular values of ``a``
+
+    .. note::
+
+      This procedure depends on the :mod:`LAPACK` module, and will generate a
+      compiler error if ``lapackImpl`` is ``none``.
+*/
+proc leastSquares(A: [] ?t, b: [] t, cond = -1.0) throws
+  where A.rank == 2 && b.rank == 1 && usingLAPACK && isLAPACKType(t)
+{
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1357]: Can't remember offhand where we stand on this (1) in general, (2) in LinearAlgebra module, but do we need an overload of this without `where` and give a compilerError there?

File: modules/packages/LinearAlgebra.chpl [hunk 5]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+  var matrix_order = lapack_memory_order.row_major;
+
+  var s: [0..<min((...A.shape))] real(64);
+  var rank: c_int;
+
+  var info = LAPACK.gelsd(matrix_order, workA, workB, s, rcond, rank);
+
+  // Check for errors
+  if info < 0 then
+    throw new owned IllegalArgumentError('gelsd(): Argument %i incorrect'.format(info));
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1396]: "is incorrect" ? and `-info`

File: modules/packages/LinearAlgebra.chpl [hunk 6]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+  else if info > 0 then
+    throw new owned LinearAlgebraError('gelsd(): SVD failed to converge with %i off-diagonal elements not converged to 0'.format(info));
+
+  var x1 = workB[0..<n, 0];
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1400]: I can't find what `n` is, but I am probably missing something

File: modules/packages/LinearAlgebra.chpl [hunk 7]
```diff
@@ -1331,6 +1331,100 @@ proc solve (A: [?Adom] ?eltType, b: [?bdom] eltType) {
...
+/* Generate Vandermonde matrix */
+proc vander(x: [?d], in N=0) where d.rank == 1 {
+
+  if N == 0 then N = d.size;
+
+  var resultDom = {d.dim(0), 0..<N};
+  var result: [resultDom] x.eltType;
+
+  forall i in resultDom.dim(0) {
+    for j in resultDom.dim(1) by -1 {
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1420]: I don't know the math side of this, but I can't see why we are iterating in reverse and not just doing `forall (i,j) in resultDom` (If we can it'll have performance benefits if/when we use distributed domains/arrays)

File: modules/packages/LinearAlgebra.chpl [hunk 8]
```diff
@@ -1722,6 +1815,10 @@ proc type _array.rank param {
   return x.rank;
 }
 
+//
+// Type helperr
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1819]: rrrr

File: modules/packages/LinearAlgebra.chpl (hunk 6 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl:1400]: 1370: `const (m, n) = A.shape;`

File: modules/packages/LinearAlgebra.chpl (hunk 7 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl:1420]: I was seeing an issue when zipping these before, but I plan on whittling down to a clean reproducer and filing as an issue.

File: modules/packages/LinearAlgebra.chpl (hunk 4 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl:1357]: We're not that consistent for (1) or (2). Do you have a preference as the reviewer? Rather than an overload, I would lean towards checking ranks in the function body and emitting a more helpful error message.

File: modules/packages/LinearAlgebra.chpl (hunk 3 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl:1347]: Indeed. Do you have any suggestions?

File: modules/packages/LinearAlgebra.chpl (hunk 2 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl:1346]: I got the bullets to look good, but then realized this is not needed until we support 2D `b` argument, so I'm simplifying the text instead.

File: modules/packages/LinearAlgebra.chpl (hunk 3 above)
Comment (e-kayrakli) [reply modules/packages/LinearAlgebra.chpl]: This is probably the curse of the LinearAlgebra module in general. It is just something that we need to live with, I think.

I was mostly being grumpy as it took me few seconds to understand what you are trying to say there :(

File: modules/packages/LinearAlgebra.chpl (hunk 4 above)
Comment (e-kayrakli) [reply modules/packages/LinearAlgebra.chpl:1361]: I prefer to have an error message coming from the module rather than a resolution error coming from the compiler. Have no preference about how it is done, though.

**Comment by ben-albrecht** [comment]
@e-kayrakli - I have (belatedly) addressed all your feedback. I think this is ready for re-review when you get a chance.

Testing:

- [ ] `start_test test/library/packages/{LinearAlgebra, LAPACK}` on Cray XC

**Comment by ben-albrecht** [comment]
@e-kayrakli - OK, testing passed now. Ready for re-review

File: modules/packages/LinearAlgebra.chpl [hunk 9]
```diff
@@ -1882,7 +1984,7 @@ proc svd(A: [?Adom] ?t) throws
 proc jacobi(A: [?Adom] ?eltType, ref X: [?Xdom] eltType,
             b: [Xdom] eltType, tol = 0.0001, maxiter = 1000) {
   if Adom.rank != 2 || X.rank != 1 || b.rank != 1 then
-    halt("Wrong shape of input matrix or vector");
+    compilerError("Wrong shape of input matrix or vector");
```
Comment (e-kayrakli) [inline modules/packages/LinearAlgebra.chpl:1987]: This is not part of this PR, but this error message should say that it is coming from `jacobi`

**Review by e-kayrakli — APPROVED** [review]

File: modules/packages/LinearAlgebra.chpl (hunk 9 above)
Comment (ben-albrecht) [reply modules/packages/LinearAlgebra.chpl]: Yeah, we are not consistent here either. I've updated the jacobi cases for this PR, but think updating all cases deserves a separate effort.
